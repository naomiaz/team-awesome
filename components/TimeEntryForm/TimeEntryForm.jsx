import React from 'react';
// import PropTypes from 'prop-types';

import './TimeEntryForm.scss';

const TimeEntryForm = () => (
  <div className="row">
    <form className="time-entry">
      <div className="time-entry-wrapper">
        <div className="time-entry-employer">
          <label className="time-entry__label" htmlFor="employer">
            Employer
            <select className="time-entry__input" name="employer">
              <option value="Port of Rotterdam">
                Port of Rotterdam
              </option>
              <option value="Saab">
                Saab
              </option>
              <option value="Mercedes">
                Mercedes
              </option>
              <option value="Audi">
                Audi
              </option>
            </select>
          </label>
        </div>

        <div className="time-entry-activity">
          <label className="time-entry__label" htmlFor="activity">
            Activity
            <select className="time-entry__input" name="activity" id="activity">
              <option value="Design">
                Design
              </option>
              <option value="Saab">
                Saab
              </option>
              <option value="Mercedes">
                Mercedes
              </option>
              <option value="Audi">
                Audi
              </option>
            </select>
          </label>
        </div>

        <div className="time-entry-date">
          <label className="time-entry__label" htmlFor="date">
            Date
            <input type="date" name="date" className="time-entry__input" />
          </label>
        </div>

        <div className="time-entry-time">
          <div className="time-entry-time__from">
            <label className="time-entry__label" htmlFor="time-from">
              From
              <input type="time" name="time-from" className="time-entry__input" />
            </label>
          </div>
          <div className="time-entry-time__to">
            <label className="time-entry__label" htmlFor="time-to">
              To
              <input type="time" name="time-to" className="time-entry__input" />
            </label>
          </div>
        </div>
      </div>

      <button type="button" name="button" value="Add" className="time-entry__button">
        Add
      </button>
    </form>
  </div>
);

export default TimeEntryForm;
