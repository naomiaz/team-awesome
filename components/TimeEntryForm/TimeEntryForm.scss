@import "../../shared/styles/main";

.time-entry {
  background-color: $color-monochrome-2;
  border: 1px solid $color-monochrome-6;
  border-radius: $border-radius-primary;
  display: flex;
  flex-flow: column;
  margin-top: 24px;
  padding: 16px;

  &__label {
    color: $color-monochrome-4;
    display: block;
    font-size: 12px;
    font-weight: 600;
    text-transform: uppercase;
  }

  &__input {
    background-image: linear-gradient($color-monochrome-1, $color-monochrome-2);
    border: 1px solid $color-monochrome-6;
    border-radius: $border-radius-primary;
    font-family: $font-primary;
    font-size: 14px;
    font-weight: 600;
    height: 36px;
    margin: 10px 0;
    padding: 0 15px;
    width: 100%;
  }

  &__button {
    background-color: $color-accent-alt-5;
    border: none;
    border-radius: $border-radius-primary;
    color: $color-monochrome-1;
    font-size: 14px;
    font-weight: 400;
    margin: 15px 0;
    padding: 10px 0;
  }

  &-time {
    display: flex;
    flex-flow: row;
    justify-content: space-between;

    &__from,
    &__to {
      flex-basis: 49%;
    }
  }
}

@include tablet {
  .time-entry {
    display: flex;
    flex-flow: row;
    justify-content: space-between;

    &-wrapper {
      display: flex;
      flex-basis: 75%;
      justify-content: space-between;
    }

    &-employer,
    &-activity {
      flex-basis: 245px;
    }

    &-date {
      flex-basis: 130px;
    }

    &-time {
      flex-basis: 64px;
    }

    &__button {
      width: 90px;
    }
  }

}
