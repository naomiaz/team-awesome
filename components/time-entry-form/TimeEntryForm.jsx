import React from 'react';
import PropTypes from 'prop-types';

import './time-entry-form.scss';

const TimeEntryForm = ({
  isFormVisible, formVisible, handleChange, onEntrySubmit
}) => (
  <section className="row">
    <button
      className={`time-entry__button-new${isFormVisible ? '--hidden' : '--visible'}`}
      onClick={formVisible}
      type="button"
    >
      <svg className="time-entry__icon--open" />
      New time entry
    </button>

    <form
      className={`time-entry ${isFormVisible ? ' time-entry--visible' : ' time-entry--hidden'}`}
      onSubmit={onEntrySubmit}
    >
      <div className="time-entry-wrapper">
        <button
          className="time-entry__button-close"
          onClick={formVisible}
          type="button"
        >
          <svg className="time-entry__icon--close" />
        </button>

        <div className="time-entry-client">
          <label
            className="time-entry__label"
            htmlFor="client"
          >
            Client
            <select
              className="time-entry__input"
              id="client"
              name="client"
              onChange={handleChange}
            >
              <option value="Port of Rotterdam">
                Port of Rotterdam
              </option>
              <option value="Saab">
                Saab
              </option>
              <option value="Mercedes">
                Mercedes
              </option>
              <option value="Audi">
                Audi
              </option>
            </select>
          </label>
        </div>

        <div className="time-entry-activity">
          <label
            className="time-entry__label"
            htmlFor="activity"
          >
            Activity
            <select
              className="time-entry__input"
              id="activity"
              name="activity"
              onChange={handleChange}
            >
              <option value="Design">
                Design
              </option>
              <option value="Saab">
                Saab
              </option>
              <option value="Mercedes">
                Mercedes
              </option>
              <option value="Audi">
                Audi
              </option>
            </select>
          </label>
        </div>

        <div className="time-entry-date">
          <label
            className="time-entry__label"
            htmlFor="date"
          >
            Date
            <input
              className="time-entry__input"
              id="date"
              name="date"
              onChange={handleChange}
              type="date"
            />
          </label>
        </div>

        <div className="time-entry-time">
          <div className="time-entry-time__from">
            <label
              className="time-entry__label"
              htmlFor="time-from"
            >
              From
              <input
                className="time-entry__input"
                defaultValue="08:00"
                id="time-from"
                max="18:00"
                min="08:00"
                name="time-from"
                onChange={handleChange}
                type="time"
              />
            </label>
          </div>
          <div className="time-entry-time__to">
            <label
              className="time-entry__label"
              htmlFor="time-to"
            >
              To
              <input
                className="time-entry__input"
                defaultValue="18:00"
                id="time-to"
                max="20:00"
                min="08:00"
                name="time-to"
                onChange={handleChange}
                type="time"
              />
            </label>
          </div>
        </div>
      </div>

      <button
        className="time-entry__button-add"
        name="button"
        // onClick={() => this.props.onEntrySubmit(this.state)}
        type="submit"
        value="Add"
      >
        Add
      </button>
    </form>
  </section>
);

TimeEntryForm.propTypes = {
  isFormVisible: PropTypes.bool.isRequired,
  formVisible: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  onEntrySubmit: PropTypes.func.isRequired
};

// class TimeEntryForm extends React.Component {
//   static propTypes = {
//     onEntrySubmit: PropTypes.func.isRequired
//   }

  // state = {
  //   isFormVisible: false
  // };

  // formVisible = () => {
  //   this.setState(prevState => ({
  //     isFormVisible: !prevState.isFormVisible
  //   }));
  // }
  //
  // handleChange = ({ target }) => {
  //   // handleChange vult de state met de waardes uit het formulier (setState(this.state))
  //   this.setState(prevState => ({
  //     ...prevState,
  //     [target.name]: target.value
  //   }));
  // }

  // handleSubmit = (event) => {
  //   event.preventDefault();
  //   onEntrySubmit(this.state);
  //   // handleSubmit op het form die de state terugstuurt naar boven -> map() naar newEntry{}
  // }

//   render() {
//     const { isFormVisible } = this.state;
//     const { onEntrySubmit } = this.props;
//     return (
//       <section className="row">
//         <button
//           className={`time-entry__button-new${isFormVisible ? '--hidden' : '--visible'}`}
//           onClick={this.formVisible}
//           type="button"
//         >
//           <svg className="time-entry__icon--open" />
//           New time entry
//         </button>
//
//         <form
//           className={`time-entry ${isFormVisible ? ' time-entry--visible' : ' time-entry--hidden'}`}
//           onSubmit={onEntrySubmit}
//         >
//           <div className="time-entry-wrapper">
//             <button
//               className="time-entry__button-close"
//               onClick={this.formVisible}
//               type="button"
//             >
//               <svg className="time-entry__icon--close" />
//             </button>
//
//             <div className="time-entry-client">
//               <label
//                 className="time-entry__label"
//                 htmlFor="client"
//               >
//                 Client
//                 <select
//                   className="time-entry__input"
//                   id="client"
//                   name="client"
//                   onChange={this.handleChange}
//                 >
//                   <option value="Port of Rotterdam">
//                     Port of Rotterdam
//                   </option>
//                   <option value="Saab">
//                     Saab
//                   </option>
//                   <option value="Mercedes">
//                     Mercedes
//                   </option>
//                   <option value="Audi">
//                     Audi
//                   </option>
//                 </select>
//               </label>
//             </div>
//
//             <div className="time-entry-activity">
//               <label
//                 className="time-entry__label"
//                 htmlFor="activity"
//               >
//                 Activity
//                 <select
//                   className="time-entry__input"
//                   id="activity"
//                   name="activity"
//                   onChange={this.handleChange}
//                 >
//                   <option value="Design">
//                     Design
//                   </option>
//                   <option value="Saab">
//                     Saab
//                   </option>
//                   <option value="Mercedes">
//                     Mercedes
//                   </option>
//                   <option value="Audi">
//                     Audi
//                   </option>
//                 </select>
//               </label>
//             </div>
//
//             <div className="time-entry-date">
//               <label
//                 className="time-entry__label"
//                 htmlFor="date"
//               >
//                 Date
//                 <input
//                   className="time-entry__input"
//                   id="date"
//                   name="date"
//                   onChange={this.handleChange}
//                   type="date"
//                 />
//               </label>
//             </div>
//
//             <div className="time-entry-time">
//               <div className="time-entry-time__from">
//                 <label
//                   className="time-entry__label"
//                   htmlFor="time-from"
//                 >
//                   From
//                   <input
//                     className="time-entry__input"
//                     defaultValue="08:00"
//                     id="time-from"
//                     max="18:00"
//                     min="08:00"
//                     name="time-from"
//                     onChange={this.handleChange}
//                     type="time"
//                   />
//                 </label>
//               </div>
//               <div className="time-entry-time__to">
//                 <label
//                   className="time-entry__label"
//                   htmlFor="time-to"
//                 >
//                   To
//                   <input
//                     className="time-entry__input"
//                     defaultValue="18:00"
//                     id="time-to"
//                     max="20:00"
//                     min="08:00"
//                     name="time-to"
//                     onChange={this.handleChange}
//                     type="time"
//                   />
//                 </label>
//               </div>
//             </div>
//           </div>
//
//           <button
//             className="time-entry__button-add"
//             name="button"
//             // onClick={() => this.props.onEntrySubmit(this.state)}
//             type="submit"
//             value="Add"
//           >
//             Add
//           </button>
//         </form>
//       </section>
//     );
//   }
// }

export default TimeEntryForm;
