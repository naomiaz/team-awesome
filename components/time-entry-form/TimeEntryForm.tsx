import React from 'react';

import Button from '../../shared/components/button/Button';
import SelectBox from '../../shared/components/select-box/SelectBox';
import InputField from '../../shared/components/input-field/InputField';

import { TimeEntryModel } from '../../ducks/time-entries';
import { ClientNameModel } from '../../ducks/clients';
import { convertTimeToIso, convertDateToIso, createIsoString } from '../../services/date-time/date-time';

import './time-entry-form.scss';

export interface TimeEntryFormProps {
  clientNames: ClientNameModel[];
  isFormSaving: boolean;
  isFormVisible: boolean;
  onEntrySubmit: (newTimeEntryFormatted: TimeEntryModel) => void;
  onToggleFormVisibility: (isFormVisible: boolean) => void;
}

export interface TimeEntryFormState {
  timeEntry: TimeEntryModel;
  validity: { date: boolean, timeFrom: boolean, timeTo: boolean };
}

class TimeEntryForm extends React.Component <TimeEntryFormProps, TimeEntryFormState> {
  private defaultState: TimeEntryFormState = {
    timeEntry: {
      activity: 'Design',
      clientId: '1',
      date: '',
      timeFrom: '',
      timeTo: ''
    },
    validity: {
      date: true,
      timeFrom: true,
      timeTo: true
    }
  };

  private formElement = React.createRef<HTMLFormElement>();

  readonly state: TimeEntryFormState = this.defaultState;

  handleFormVisibility = () => {
    this.props.onToggleFormVisibility(!this.props.isFormVisible);
  };

  handleBlur = ({ target }) => {
    this.setState((prevState: TimeEntryFormState) => ({
      validity: {
        ...prevState.validity,
        [target.name]: target.validity.valid
      }
    }));
  };

  handleChange = ({ target }) => {
    this.setState((prevState: TimeEntryFormState) => ({
      timeEntry: {
        ...prevState.timeEntry,
        [target.name]: target.value
      }
    }));
  };

  checkFormValidation = () => (
    // First check if the formElement exists
    // Then loop over each formElement and check its validity -> .every() returns boolean
    this.formElement.current && Array
      .from(this.formElement.current.elements)
      .every((input: HTMLInputElement) => input.validity.valid)
  );

  handleSubmit = (event) => {
    const { timeEntry } = this.state;
    const { date, timeFrom, timeTo } = timeEntry;
    event.preventDefault();

    const newTimeEntryFormatted: TimeEntryModel = {
      ...timeEntry,
      date: convertDateToIso(date),
      timeFrom: createIsoString(convertDateToIso(date), convertTimeToIso(timeFrom)),
      timeTo: createIsoString(convertDateToIso(date), convertTimeToIso(timeTo))
    };

    if (this.checkFormValidation()) {
      this.props.onEntrySubmit(newTimeEntryFormatted);
      this.setState(() => this.defaultState);
    }
  };

  render(): React.ReactNode {
    const { timeEntry, validity } = this.state;
    const {
      clientId, activity, date, timeFrom, timeTo
    } = timeEntry;
    const { clientNames, isFormSaving, isFormVisible } = this.props;
    return (
      <section className="time-entry-form">
        <h2 className="time-entry-form__title">
          New Time Entry
        </h2>
        <Button
          className={`
            time-entry-form__button-new
            time-entry-form__button-new--${isFormVisible ? 'hidden' : 'visible'}
          `}
          disabled={null}
          onClick={this.handleFormVisibility}
          type="button"
          value="New time entry"
        />

        <form
          className={`
            time-entry-form__form-container
            time-entry-form__form-container--${isFormVisible ? 'visible' : 'hidden'}`}
          onSubmit={this.handleSubmit}
          ref={this.formElement}
        >
          <div className="time-entry-form__form-wrapper">

            <button
              className="time-entry-form__button-close"
              onClick={this.handleFormVisibility}
              type="button"
            >
              <svg className="time-entry-form__icon-close" />
            </button>

            <label
              className="time-entry-form__label"
              htmlFor="clientId"
            >
              Client
              <SelectBox
                className="
                  time-entry-form__input
                  time-entry-form__select
                "
                selectedValue={clientId}
                name="clientId"
                onChange={this.handleChange}
                options={clientNames}
              />
            </label>

            <label
              className="time-entry-form__label"
              htmlFor="activity"
            >
              Activity
              <SelectBox
                className="
                  time-entry-form__input
                  time-entry-form__select
                "
                selectedValue={activity}
                name="activity"
                onChange={this.handleChange}
                options={[
                  { title: 'Design', value: 'Design' },
                  { title: 'Development', value: 'Development' },
                  { title: 'Meeting', value: 'Meeting' },
                  { title: 'Traveling', value: 'Traveling' }
                ]}
              />
            </label>

            {/* DATE */}
            <label
              className="time-entry-form__label"
              htmlFor="date"
            >
              Date
              <InputField
                className={`
                  time-entry-form__input
                  time-entry-form__input--${validity.date ? 'valid' : 'invalid'}
                `}
                name="date"
                onBlur={this.handleBlur}
                onChange={this.handleChange}
                pattern="(0[1-9]|[1-2][0-9]|3[0-1])[-](0[1-9]|1[0-2])[-]20[0-1][0-9]"
                placeholder="DD-MM-YYYY"
                required
                type="text"
                value={date}
              />
            </label>

            <div className="time-entry-form__time-wrapper">
              <label
                className="
                  time-entry-form__label
                  time-entry-form__label--half
                  time-entry-form__time-from
                "
                htmlFor="time-from"
              >
                From
                <InputField
                  className={`
                    time-entry-form__input
                    time-entry-form__input--${validity.timeFrom ? 'valid' : 'invalid'}
                  `}
                  name="timeFrom"
                  onBlur={this.handleBlur}
                  onChange={this.handleChange}
                  pattern="(0[0-9]|1[0-9]|2[0-3])[.](0[0-9]|[1-5][0-9])"
                  placeholder="HH.MM"
                  required
                  type="text"
                  value={timeFrom}
                />
              </label>

              <label
                className="
                  time-entry-form__label
                  time-entry-form__time-to
                "
                htmlFor="time-to"
              >
                To
                <InputField
                  className={`
                    time-entry-form__input
                    time-entry-form__input--${validity.timeTo ? 'valid' : 'invalid'}
                  `}
                  name="timeTo"
                  onBlur={this.handleBlur}
                  onChange={this.handleChange}
                  pattern="(0[0-9]|1[0-9]|2[0-3])[.](0[0-9]|[1-5][0-9])"
                  placeholder="HH.MM"
                  required
                  type="text"
                  value={timeTo}
                />
              </label>
            </div>
          </div>

          <Button
            // Disable the Add button if the form is saving or if the form isn't valid (false)
            // this.checkFormValidation() is invoked each time the state updates after blurring ->
            // It will return return true/become enabled once all items are valid
            disabled={isFormSaving || !this.checkFormValidation()}
            className="time-entry-form__button-add"
            onClick={null}
            type="submit"
            value="Add"
          >
            Add
          </Button>
        </form>
      </section>
    );
  }
}

export default TimeEntryForm;
