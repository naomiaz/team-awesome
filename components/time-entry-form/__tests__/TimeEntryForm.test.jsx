import React from 'react';
import { shallow } from 'enzyme';

import TimeEntryForm from '../TimeEntryForm';

test('is the initial state loaded in the component', () => {
  const timeEntryForm = shallow(<TimeEntryForm />);

  expect(timeEntryForm.state()).toEqual({
    timeEntry: {
      client: 'Port of Rotterdam',
      activity: 'Design',
      date: '',
      timeFrom: '',
      timeTo: ''
    },
    validity: {
      date: true,
      timeFrom: true,
      timeTo: true
    }
  });
});
