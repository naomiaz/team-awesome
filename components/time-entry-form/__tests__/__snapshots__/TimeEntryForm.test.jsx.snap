// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="row"
>
  <h2
    className="time-entry__title"
  >
    New Time Entry
  </h2>
  <button
    className="btn time-entry__button-new--visible"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="time-entry__icon--open"
    />
    New time entry
  </button>
  <form
    className="time-entry time-entry--hidden"
    onSubmit={[Function]}
  >
    <div
      className="time-entry-wrapper"
    >
      <button
        className="time-entry__button-close"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="time-entry__icon--close"
        />
      </button>
      <div
        className="time-entry-client"
      >
        <label
          className="time-entry__label"
          htmlFor="client"
        >
          Client
          <select
            className="time-entry__input"
            id="client"
            name="client"
            onChange={[Function]}
            value="Port of Rotterdam"
          >
            <option
              value="Port of Rotterdam"
            >
              Port of Rotterdam
            </option>
            <option
              value="Saab"
            >
              Saab
            </option>
            <option
              value="Mercedes"
            >
              Mercedes
            </option>
            <option
              value="Audi"
            >
              Audi
            </option>
          </select>
        </label>
      </div>
      <div
        className="time-entry-activity"
      >
        <label
          className="time-entry__label"
          htmlFor="activity"
        >
          Activity
          <select
            className="time-entry__input"
            id="activity"
            name="activity"
            onChange={[Function]}
            value="Design"
          >
            <option
              value="Design"
            >
              Design
            </option>
            <option
              value="Saab"
            >
              Saab
            </option>
            <option
              value="Mercedes"
            >
              Mercedes
            </option>
            <option
              value="Audi"
            >
              Audi
            </option>
          </select>
        </label>
      </div>
      <div
        className="time-entry-date"
      >
        <label
          className="time-entry__label"
          htmlFor="date"
        >
          Date
          <input
            className="
                    time-entry__input
                    time-entry__input--valid
                  "
            id="date"
            name="date"
            onBlur={[Function]}
            onChange={[Function]}
            pattern="(0[1-9]|[1-2][0-9]|3[0-1])[-](0[1-9]|1[0-2])[-]20[0-1][0-9]"
            placeholder="DD-MM-YYYY"
            required={true}
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        className="time-entry-time"
      >
        <div
          className="time-entry-time__from"
        >
          <label
            className="time-entry__label"
            htmlFor="time-from"
          >
            From
            <input
              className="
                      time-entry__input
                      time-entry__input--valid
                    "
              id="time-from"
              name="timeFrom"
              onBlur={[Function]}
              onChange={[Function]}
              pattern="(0[0-9]|1[0-9]|2[0-3])[.](0[0-9]|[1-5][0-9])"
              placeholder="HH.MM"
              required={true}
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          className="time-entry-time__to"
        >
          <label
            className="time-entry__label"
            htmlFor="time-to"
          >
            To
            <input
              className="
                      time-entry__input
                      time-entry__input--valid
                    "
              id="time-to"
              name="timeTo"
              onBlur={[Function]}
              onChange={[Function]}
              pattern="(0[0-9]|1[0-9]|2[0-3])[.](0[0-9]|[1-5][0-9])"
              placeholder="HH.MM"
              required={true}
              type="text"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
    <button
      className="btn time-entry__button-add"
      disabled={true}
      name="button"
      type="submit"
      value="Add"
    >
      Add
    </button>
  </form>
</section>
`;
